src/common/src/common-string-private.h

```diff
@@ -20,75 +20,537 @@
+ * @param new_append Pointer to an uninitialized mcommon_string_append_t
+ * @param string String allocated with mcommon_string_new
+ * @param max_len Maximum allowed length for the resulting string, in bytes. Must be less than UINT32_MAX.
+ *
+ * The mcommon_string_append_t does not need to be deallocated. It is no longer usable if the underlying
+ * mcommon_string_t is freed.
+ *
+ * If the string was already over maximum length, it will not be modified. All append operations are guaranteed not to
+ * lengthen the string beyond max_len. Truncations are guaranteed to happen at UTF-8 code point boundaries.
+ */
 static BSON_INLINE void
-mcommon_string_append (mcommon_string_t *string, const char *str)
+mcommon_string_append_init_with_limit (mcommon_string_append_t *new_append, mcommon_string_t *string, uint32_t max_len)
 {
-   BEGIN_IGNORE_DEPRECATIONS
-   bson_string_append (string, str);
-   END_IGNORE_DEPRECATIONS
+   BSON_ASSERT_PARAM (new_append);
+   BSON_ASSERT_PARAM (string);
+   BSON_ASSERT (max_len < UINT32_MAX);
+   new_append->_string = string;
+   new_append->_max_len = max_len;
+   new_append->_max_len_exceeded = false;
 }
+
+/**
+ * @brief Begin appending to an mcommon_string_t
+ * @param new_append Pointer to an uninitialized mcommon_string_append_t
+ * @param string String allocated with mcommon_string_new
+ *
+ * The mcommon_string_append_t does not need to be deallocated. It is no longer usable if the underlying
+ * mcommon_string_t is freed.
+ *
+ * The maximum string length will be set to the largest representable by the data type, UINT32_MAX - 1.
+ */
 static BSON_INLINE void
-mcommon_string_append_c (mcommon_string_t *string, char str)
+mcommon_string_append_init (mcommon_string_append_t *new_append, mcommon_string_t *string)
 {
-   BEGIN_IGNORE_DEPRECATIONS
-   bson_string_append_c (string, str);
-   END_IGNORE_DEPRECATIONS
+   mcommon_string_append_init_with_limit (new_append, string, UINT32_MAX - 1u);
 }
+
+/**
+ * @brief Begin appending to a new empty mcommon_string_t with default initial capacity and maximum length
+ * @param new_append Pointer to an uninitialized mcommon_string_append_t
+ *
+ * Allocates a new mcommon_string_t, which will need to be deallocated by the caller.
+ * The mcommon_string_append_t itself does not need to be deallocated.
+ *
+ * The maximum string length will be set to the largest representable by the data type, UINT32_MAX - 1.
+ * The new string will be allocated with a small default capacity.
+ *
+ * This method is intended to be the most convenient way to start growing a string. If a reasonable guess
+ * can be made about the final size of the string, it's better to call mcommon_string_new_with_capacity()
+ * and mcommon_string_append_init() directly.
+ */
 static BSON_INLINE void
-mcommon_string_append_unichar (mcommon_string_t *string, bson_unichar_t unichar)
+mcommon_string_append_new (mcommon_string_append_t *new_append)
```

Comment:
My intent with 'new' here was to focus on the new mcommon_string_t being created. A more verbose name might be mcommon_string_append_init_with_new_string, but at the time this seemed like a better naming compromise. I'm not sure the _init_with_string name serves well, because I'd expect nearly all of the mcommon_string_append constructors to take a mcommon_string_t as argument. Constructing a mcommon_string_t at the same time as the mcommon_string_append_t is a convenience, not a built-in feature of the string_append. I'll think about a clearer naming for these.